### ISSUE TITLE
[REVIEW]: COLNA: A Python Package to Analyze Complex, Linear Networks

### ISSUE THREAD

**Issue by whedon** [openjournals/joss-reviews#2073]
**Submitting author:** @pstrk (<a href="http://orcid.org/0000-0002-7389-5592">Pascal Stark</a>)
**Repository:** <a href="https://github.com/IBM/complex-linear-network-analyzer" target ="_blank">https://github.com/IBM/complex-linear-network-analyzer</a>
**Version:** v1.0
**Editor:** @arfon
**Reviewer:** @brandondube, @DiegoJRR
**Archive:** <a href="https://doi.org/10.5281/zenodo.3686634" target="_blank">10.5281/zenodo.3686634</a>

## Status

[![status](https://joss.theoj.org/papers/6e5d2d8fa96c3aa8d960d78353960bae/status.svg)](https://joss.theoj.org/papers/6e5d2d8fa96c3aa8d960d78353960bae)

Status badge code:

```
HTML: <a href="https://joss.theoj.org/papers/6e5d2d8fa96c3aa8d960d78353960bae"><img src="https://joss.theoj.org/papers/6e5d2d8fa96c3aa8d960d78353960bae/status.svg"></a>
Markdown: [![status](https://joss.theoj.org/papers/6e5d2d8fa96c3aa8d960d78353960bae/status.svg)](https://joss.theoj.org/papers/6e5d2d8fa96c3aa8d960d78353960bae)
```
**Reviewers and authors**:

Please avoid lengthy details of difficulties in the review thread. Instead, please create a new issue in the <a href="https://github.com/IBM/complex-linear-network-analyzer" target="_blank">target repository</a> and link to those issues (especially acceptance-blockers) by leaving comments in the review thread below. (For completists: if the target issue tracker is also on GitHub, linking the review thread in the issue or vice versa will create corresponding breadcrumb trails in the link target.)

## Reviewer instructions & questions

@brandondube & @DiegoJRR, please carry out your review in this issue by updating the checklist below. If you cannot edit the checklist please:
1. Make sure you're logged in to your GitHub account
2. Be sure to accept the invite at this URL: https://github.com/openjournals/joss-reviews/invitations

The reviewer guidelines are available here: https://joss.readthedocs.io/en/latest/reviewer_guidelines.html. Any questions/concerns please let @arfon know.

✨ **Please try and complete your review in the next two weeks** ✨ 

### Conflict of interest

- [x] I confirm that I have read the [JOSS conflict of interest (COI) policy](https://github.com/openjournals/joss/blob/master/COI.md) and that: I have no COIs with reviewing this work or that any perceived COIs have been waived by JOSS for the purpose of this review.

### Code of Conduct

- [x] I confirm that I read and will adhere to the [JOSS code of conduct](https://joss.theoj.org/about#code_of_conduct).

### General checks

- [x] **Repository:** Is the source code for this software available at the <a target="_blank" href="https://github.com/IBM/complex-linear-network-analyzer">repository url</a>?
- [x] **License:** Does the repository contain a plain-text LICENSE file with the contents of an [OSI approved](https://opensource.org/licenses/alphabetical) software license?
- [x] **Contribution and authorship:** Has the submitting author (@pstrk) made major contributions to the software? Does the full list of paper authors seem appropriate and complete?

### Functionality

- [x] **Installation:** Does installation proceed as outlined in the documentation?
- [x] **Functionality:** Have the functional claims of the software been confirmed?
- [x] **Performance:** If there are any performance claims of the software, have they been confirmed? (If there are no claims, please check off this item.)

### Documentation

- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **Installation instructions:** Is there a clearly-stated list of dependencies? Ideally these should be handled with an automated package management solution.
- [x] **Example usage:** Do the authors include examples of how to use the software (ideally to solve real-world analysis problems).
- [x] **Functionality documentation:** Is the core functionality of the software documented to a satisfactory level (e.g., API method documentation)?
- [x] **Automated tests:** Are there automated tests or manual steps described so that the functionality of the software can be verified?
- [x] **Community guidelines:** Are there clear guidelines for third parties wishing to 1) Contribute to the software 2) Report issues or problems with the software 3) Seek support

### Software paper

- [x] **Summary:** Has a clear description of the high-level functionality and purpose of the software for a diverse, non-specialist audience been provided?
- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **State of the field:** Do the authors describe how this software compares to other commonly-used packages?
- [x] **Quality of writing:** Is the paper well written (i.e., it does not require editing for structure, language, or writing quality)?
- [x] **References:** Is the list of references complete, and is everything cited appropriately that should be cited (e.g., papers, datasets, software)? Do references in the text use the proper [citation syntax]( https://rmarkdown.rstudio.com/authoring_bibliographies_and_citations.html#citation_syntax)?

### Conflict of interest

- [x] I confirm that I have read the [JOSS conflict of interest (COI) policy](https://github.com/openjournals/joss/blob/master/COI.md) and that: I have no COIs with reviewing this work or that any perceived COIs have been waived by JOSS for the purpose of this review.

### Code of Conduct

- [x] I confirm that I read and will adhere to the [JOSS code of conduct](https://joss.theoj.org/about#code_of_conduct).

### General checks

- [x] **Repository:** Is the source code for this software available at the <a target="_blank" href="https://github.com/IBM/complex-linear-network-analyzer">repository url</a>?
- [x] **License:** Does the repository contain a plain-text LICENSE file with the contents of an [OSI approved](https://opensource.org/licenses/alphabetical) software license?
- [x] **Contribution and authorship:** Has the submitting author (@pstrk) made major contributions to the software? Does the full list of paper authors seem appropriate and complete?

### Functionality

- [x] **Installation:** Does installation proceed as outlined in the documentation?
- [x] **Functionality:** Have the functional claims of the software been confirmed?
- [x] **Performance:** If there are any performance claims of the software, have they been confirmed? (If there are no claims, please check off this item.)

### Documentation

- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **Installation instructions:** Is there a clearly-stated list of dependencies? Ideally these should be handled with an automated package management solution.
- [x] **Example usage:** Do the authors include examples of how to use the software (ideally to solve real-world analysis problems).
- [x] **Functionality documentation:** Is the core functionality of the software documented to a satisfactory level (e.g., API method documentation)?
- [x] **Automated tests:** Are there automated tests or manual steps described so that the functionality of the software can be verified?
- [x] **Community guidelines:** Are there clear guidelines for third parties wishing to 1) Contribute to the software 2) Report issues or problems with the software 3) Seek support

### Software paper

- [x] **Summary:** Has a clear description of the high-level functionality and purpose of the software for a diverse, non-specialist audience been provided?
- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **State of the field:** Do the authors describe how this software compares to other commonly-used packages?
- [x] **Quality of writing:** Is the paper well written (i.e., it does not require editing for structure, language, or writing quality)?
- [x] **References:** Is the list of references complete, and is everything cited appropriately that should be cited (e.g., papers, datasets, software)? Do references in the text use the proper [citation syntax]( https://rmarkdown.rstudio.com/authoring_bibliographies_and_citations.html#citation_syntax)?

**Comment by whedon** [comment]
Hello human, I'm @whedon, a robot that can help you with some common editorial tasks. @brandondube, @DiegoJRR it looks like you're currently assigned to review this paper :tada:.

**:star: Important :star:**

If you haven't already, you should seriously consider unsubscribing from GitHub notifications for this (https://github.com/openjournals/joss-reviews) repository. As a reviewer, you're probably currently watching this repository which means for GitHub's default behaviour you will receive notifications (emails) for all reviews 😿

To fix this do the following two things:

1. Set yourself as 'Not watching' https://github.com/openjournals/joss-reviews:

![watching](https://cloud.githubusercontent.com/assets/4483/20250593/64d7ce48-a9de-11e6-9225-d3dfb3e48e68.png)

2. You may also like to change your default settings for this watching repositories in your GitHub profile here: https://github.com/settings/notifications

![notifications](https://cloud.githubusercontent.com/assets/4483/20250604/79eb3144-a9de-11e6-9e51-0fe940966689.png)


For a list of things I can do to help you, just type:

```
@whedon commands
```

For example, to regenerate the paper pdf after making changes in the paper's md or bib files, type:

```
@whedon generate pdf
```

**Comment by whedon** [comment]
```
Reference check summary:

OK DOIs

- 10.1007/978-3-642-10503-6 is OK
- 10.1364/JOSAB.29.000896 is OK
- 10.1002/lpor.201100017 is OK
- 10.1038/ncomms4541 is OK

MISSING DOIs

- None

INVALID DOIs

- None
```

**Comment by whedon** [comment]
[ :point_right: Check article proof :page_facing_up: :point_left: ](https://github.com/openjournals/joss-papers/blob/joss.02073/joss.02073/10.21105.joss.02073.pdf)

**Comment by arfon** [comment]
@brandondube, @DiegoJRR - please carry out your review in this issue by updating the checklist above and giving feedback in this issue. The reviewer guidelines are available here: https://joss.readthedocs.io/en/latest/reviewer_guidelines.html

Any questions/concerns please let me know.

**Comment by brandondube** [comment]
Hello,

I am working the review - so far so good.  Some early comments:

- Lorenz, the second author, does not show up as a contributor on the GitHub repository.  Is this an error?  Or are their contributions made in a different way?

- the paper includes a nice statement of need, but I did not find a "what can I do with this" (application level, not technical) in the first few pages of the docs.  Perhaps this could be added, something like "these features can be used to...", where "to" isn't analyze a network, but, say, design a PIC.

- I see there are tests, but there is no coverage report.  I ran pytest with pytest-cov locally, and got 73% coverage.  Perhaps a statement of test coverage could be added to the repository, or a coveralls badge if the authors are interested in that sort of bling.

- I tried running a few examples.  There are various errors, some "cosmetic" in nature, others seem to disrupt the core functionality.  -- see [here](https://gist.github.com/brandondube/417b0c808521e94be256a23d389038f5).  For reproducibility, numpy 1.18.1, scipy 1.3.2, win10/amd64, numpy and scipy from conda.

- continuing with examples, the "small recurrent network" one seems to expose a performance issue, or just a lack of understanding on my part.

- - It begins at about 4,000 paths per second, and now 4 minutes later it is at about 100 paths per second and is using a hefty 18GB of RAM.  Is this expected?  Perhaps add a note in the examples that this one is particularly resource intensive.  I stopped it at `Network evaluation in progress: 140185paths [07:24, 78.09paths/s]` to try more examples.

- - I'm running this on a fairly beefy computer (i7-9700k, 32GB of RAM) and also see it is single threaded.  Is threading deferred to the user (for example, `OMP_NUM_THREADS` and related environment variables), or is there a specific enable/disable configuration for the package?

**Comment by pstrk** [comment]
Thank you for your feedback, I will quickly answer the non-technical questions. 

> Lorenz, the second author, does not show up as a contributor on the GitHub repository. Is this an error? Or are their contributions made in a different way?

Lorenz contributed about half of the code. We developed the code on a private Github instance, before moving it to the public Github. We did not want to include the full development history, as it would include some files that we can not yet share. For simplicity I just duplicated the code locally, removed all these files and uploaded the new version using my Github account. This is the reason why Lorenz does not show up as a contributer. 

> the paper includes a nice statement of need, but I did not find a "what can I do with this" (application level, not technical) in the first few pages of the docs. Perhaps this could be added, something like "these features can be used to...", where "to" isn't analyze a network, but, say, design a PIC.

Thank you for the comment, we will add a statment of need in the documentation as well.

> I see there are tests, but there is no coverage report. I ran pytest with pytest-cov locally, and got 73% coverage. Perhaps a statement of test coverage could be added to the repository, or a coveralls badge if the authors are interested in that sort of bling.

I'll work on that as well. All visualization features are not tested, we will think about a way to verify them. 

> I tried running a few examples. There are various errors, some "cosmetic" in nature, others seem to disrupt the core functionality. -- see here. For reproducibility, numpy 1.18.1, scipy 1.3.2, win10/amd64, numpy and scipy from conda.

I suggest we move this to an issue discussion in our repository. I will try to reconstruct the errors and check all the examples.

> continuing with examples, the "small recurrent network" one seems to expose a performance issue, or just a lack of understanding on my part.

I just double checked the example in the docs and there was a small typo: Edge('a', 'b', phase=0.5, attenuation=1.95, delay=1.0) should be Edge('a', 'b', phase=0.5, attenuation=0.95, delay=1.0). This leads to gain in the loop and will not converge. Eventually the evaluation would stop (after going thorugh 1e6 paths), but this will take a long time and way to much memory in the case of such a small network. I will fix the documentation accordingly. 

The network evaluation is so far indeed single threaded as you will run it only once in our typical use case, so we can afford the single thread performance limitation. Testbenchs (used to compute output signals) do support multithreading, as we often run them multiple times with different feed dictionaries.

**Comment by brandondube** [comment]
👍 for the nontechnical topics.  For the gain issue -- perhaps you could add a 'feasibility' test that can ensure a network is not divergent (within reason).

**Comment by pstrk** [comment]
> I tried running a few examples. There are various errors, some "cosmetic" in nature, others seem to disrupt the core functionality. -- see here. For reproducibility, numpy 1.18.1, scipy 1.3.2, win10/amd64, numpy and scipy from conda.

I uploaded a new version (v1.0.2) that should solve the issues you observed. Please let me know in case it does not work.

**Comment by brandondube** [comment]
Everything looks good to me now 🚀

**Comment by arfon** [comment]
:wave: @DiegoJRR - thanks again for agreeing to review this submission. Do you have a rough ETA for completing this?

**Comment by DiegoJRR** [comment]
I'll do my best to have a review for tuesday, sorry for the delay.

**Comment by DiegoJRR** [comment]
Hi, I am almost done with the review. 

I did not find real world examples in the documentation, or some direct application for the software. This would add a lot to its usability.

 Apart from this, everything LGTM! I had no problems running the examples and tests. Excellent work

**Comment by pstrk** [comment]
Hi,
Thank you for your feedback. I added a new design example (Mach Zehnder interferometer used as an optical weight), which should link more to the real world application scenarios of COLNA. Furthermore I improved the test coverage and added a few new convenience methods. I therefore released again a new version (1.0.3). 

@arfon Do I need to do anything so that we can proceed with the publication?

**Comment by arfon** [comment]
Thanks @DiegoJRR & @brandondube!

@pstrk - At this point could you make a new release of this software that includes the changes that have resulted from this review. Then, please make an archive of the software in Zenodo/figshare/other service and update this thread with the DOI of the archive? For the Zenodo/figshare archive, please make sure that:

- The title of the archive is the same as the JOSS paper title
- That the authors of the archive are the same as the JOSS paper authors

I can then move forward with accepting the submission.

**Comment by pstrk** [comment]
Hi @arfon 
Thank you for the information. I created a new release (V1.0.4) and the corresponding [Zenodo archive](https://doi.org/10.5281/zenodo.3686634). The DOI is 10.5281/zenodo.3686634.

**Comment by arfon** [comment]
@whedon set 10.5281/zenodo.3686634 as archive

**Comment by whedon** [comment]
OK. <a href="https://doi.org/10.5281/zenodo.3686634" target="_blank">10.5281/zenodo.3686634</a> is the archive.

**Comment by arfon** [comment]
@whedon accept

**Comment by whedon** [comment]
```
Attempting dry run of processing paper acceptance...
```

**Comment by whedon** [comment]
```
Reference check summary:

OK DOIs

- 10.1007/978-3-642-10503-6 is OK
- 10.1364/JOSAB.29.000896 is OK
- 10.1002/lpor.201100017 is OK
- 10.1038/ncomms4541 is OK

MISSING DOIs

- None

INVALID DOIs

- None
```

**Comment by whedon** [comment]
Check final proof :point_right: https://github.com/openjournals/joss-papers/pull/1338

If the paper PDF and Crossref deposit XML look good in https://github.com/openjournals/joss-papers/pull/1338, then you can now move forward with accepting the submission by compiling again with the flag `deposit=true` e.g.
 ```
@whedon accept deposit=true
```

**Comment by arfon** [comment]
@pstrk - I made a couple of minor changes to the text for readability here: https://github.com/IBM/complex-linear-network-analyzer/pull/21

**Comment by pstrk** [comment]
@arfon Thank you for spotting the typos, I merged the corrected version. From my side everything looks good now.

**Comment by arfon** [comment]
@whedon accept

**Comment by whedon** [comment]
```
Attempting dry run of processing paper acceptance...
```

**Comment by whedon** [comment]
```
Reference check summary:

OK DOIs

- 10.1007/978-3-642-10503-6 is OK
- 10.1364/JOSAB.29.000896 is OK
- 10.1002/lpor.201100017 is OK
- 10.1038/ncomms4541 is OK

MISSING DOIs

- None

INVALID DOIs

- None
```

**Comment by whedon** [comment]
Check final proof :point_right: https://github.com/openjournals/joss-papers/pull/1339

If the paper PDF and Crossref deposit XML look good in https://github.com/openjournals/joss-papers/pull/1339, then you can now move forward with accepting the submission by compiling again with the flag `deposit=true` e.g.
 ```
@whedon accept deposit=true
```

**Comment by arfon** [comment]
@whedon accept deposit=true

**Comment by whedon** [comment]
```
Doing it live! Attempting automated processing of paper acceptance...
```

**Comment by whedon** [comment]
🐦🐦🐦 👉 [Tweet for this paper](https://twitter.com/JOSS_TheOJ/status/1232257874488897536) 👈 🐦🐦🐦

**Comment by whedon** [comment]
🚨🚨🚨 **THIS IS NOT A DRILL, YOU HAVE JUST ACCEPTED A PAPER INTO JOSS!** 🚨🚨🚨

 Here's what you must now do:

0. Check final PDF and Crossref metadata that was deposited :point_right: https://github.com/openjournals/joss-papers/pull/1340
1. Wait a couple of minutes to verify that the paper DOI resolves [https://doi.org/10.21105/joss.02073](https://doi.org/10.21105/joss.02073)
2. If everything looks good, then close this review issue.
3. Party like you just published a paper! 🎉🌈🦄💃👻🤘

 Any issues? notify your editorial technical team...

**Comment by arfon** [comment]
@brandondube, @DiegoJRR many thanks for your reviews here ✨ 

@pstrk - your paper is now accepted into JOSS :zap::rocket::boom:

**Comment by whedon** [comment]
:tada::tada::tada: Congratulations on your paper acceptance! :tada::tada::tada:

If you would like to include a link to your paper from your README use the following code snippets:

```
Markdown:
[![DOI](https://joss.theoj.org/papers/10.21105/joss.02073/status.svg)](https://doi.org/10.21105/joss.02073)

HTML:
<a style="border-width:0" href="https://doi.org/10.21105/joss.02073">
  <img src="https://joss.theoj.org/papers/10.21105/joss.02073/status.svg" alt="DOI badge" >
</a>

reStructuredText:
.. image:: https://joss.theoj.org/papers/10.21105/joss.02073/status.svg
   :target: https://doi.org/10.21105/joss.02073
```
This is how it will look in your documentation:

[![DOI](https://joss.theoj.org/papers/10.21105/joss.02073/status.svg)](https://doi.org/10.21105/joss.02073)

**We need your help!**

Journal of Open Source Software is a community-run journal and relies upon volunteer effort. If you'd like to support us please consider doing either one (or both) of the the following:

- Volunteering to review for us sometime in the future. You can add your name to the reviewer list here: https://joss.theoj.org/reviewer-signup.html
- Making a small donation to support our running costs here: https://numfocus.org/donate-to-joss
